source/detail/parser/x86_64.cpp

```diff
@@ -31,6 +33,110 @@ namespace smeagle::x86_64 {
+
+  // Get the next framebase
+  std::string FramebaseAllocator::nextFramebaseFromType(st::Type *paramType) {
+    std::string framebaseStr = "framebase+" + std::to_string(framebase);
+
+    // Update the framebase for the next parameter based on the type
+    updateFramebaseFromType(paramType);
+
+    return framebaseStr;
+  }
+
+  // A RegisterAllocator can provide the next register location
+  class RegisterAllocator {
+  public:
+    RegisterAllocator();
+    std::string getRegisterString(RegisterClass regClass, st::Type *paramType);
+    std::string getRegistersString(std::pair<RegisterClass, RegisterClass> regClasses,
+                                   st::Type *paramType);
+
+  private:
+    class FramebaseAllocator fallocator;
+    std::stack<std::string> intRegisters{{"%r9", "%r8", "%rcx", "%rdx", "%rsi", "%rdi"}};
+    std::stack<std::string> sseRegisters;
+    std::optional<std::string> getNextIntRegister();
+    std::optional<std::string> getNextSseRegister();
+    int framebase = 8;
+  };
+
+  // Constructor
+  RegisterAllocator::RegisterAllocator() {
+    // Populate the sse register stack
+    for (int i = 1; i < 8; ++i) {
+      sseRegisters.push("xmm" + std::to_string(i));
+    }
+  }
+
+  // Get the next available integer register
+  std::optional<std::string> RegisterAllocator::getNextIntRegister() {
+    // If we are empty, return stack
+    if (intRegisters.empty()) return {};
+    std::string regString = intRegisters.top();
+    intRegisters.pop();
+    return regString;
+  }
+
+  // Get the next available integer register
+  std::optional<std::string> RegisterAllocator::getNextSseRegister() {
+    // If we are empty, return stack
+    if (intRegisters.empty()) return {};
+    std::string regString = sseRegisters.top();
+    sseRegisters.pop();
+    return regString;
+  }
+
+  // Given two registers, return one combined string
+  std::string RegisterAllocator::getRegistersString(
+      std::pair<RegisterClass, RegisterClass> regClasses, st::Type *paramType) {
+    std::string locA = this->getRegisterString(regClasses.first, paramType);
+    std::string locB = this->getRegisterString(regClasses.second, paramType);
+
+    // If B is empty (NO_CLASS) then return A
+    if (locB == "") {
+      return locA;
+    }
+    return locA + "|" + locB;
+  }
+
+  // Get a string location from a register class
+  std::string RegisterAllocator::getRegisterString(RegisterClass regClass, st::Type *paramType) {
+    std::optional<std::string> regString;
+
+    // If the class is memory, pass the argument on the stack
+    if (regClass == RegisterClass::NO_CLASS) regString = "";
+    if (regClass == RegisterClass::SSE) regString = this->getNextSseRegister();
+    if (regClass == RegisterClass::INTEGER) regString = this->getNextIntRegister();
+    if (regClass == RegisterClass::MEMORY) regString = std::nullopt;
+
+    // If we don't have a value, we need a framebase
+    if (!regString.has_value()) {
```

Comment:
std::optional has a built-in explicit conversion to bool, so you can just write `if(!regString)`, if you like.

Reply:
Gotcha! I chose this approach because I was concerned about distinguishing between an empty string and `std::nullopt`. In Python a not (!) will evaluate to true given either. It's maybe not the case in C++, but I chose to use `has_value` because it more explicitly states that we are concerned about it not being a `std::nullopt`.  I also had a question on Mattermost about doing:

```cpp
std::string result = regString.value_or(fallocator.nextFramebaseFromType(paramType));
```
In Python the second statement in an OR would only be evaluated if the first was False, but I'm not sure it's the same here! We obviously wouldn't want to generate that second value if it changes the state of the framebase allocator.

Reply:
Yeah, C++ has no "truthy" values like that. Every non-built-in type has to have an explicitly overloaded `operator!` in order to be used like that-- and most don't. The more verbose `has_value` is perfectly fine- I just wanted to point out that there is a terser version.

>In Python the second statement in an OR would only be evaluated if the first was False, but I'm not sure it's the same here!

In this case, no because the thing that would be used if the `optional` were empty is being passed as a function argument which is evaluated _before_ the function call. As such, it has to be a valid expression. C++ does have Boolean short-circuiting when using the logic operators, though. A common idiom for using an `optional` is

    if(x && x.foo())